package/MDAnalysis/analysis/MSD.py

```diff
@@ -0,0 +1,468 @@
+                 max_data=False):
+        self.universe = universe
+        self.select_list = select_list
+        if type(select_list) is str:
+            self.select_list = select_list.split(',')
+        print(self.select_list)
+        self.t0 = int(t0)
+        self.tf = int(tf)
+        self.dt_restart = int(dt_restart)
+        self.out_name = out_name
+        # handling short simulation cases
+        n_frames = int(tf) - int(t0) + 1
+        if int(len_msd) > int(n_frames):
+            len_msd = int(n_frames)
+        self.len_msd = int(len_msd)
+        self.write_output = write_output
+        self.max_data = max_data
+
+    def _select(self, frame):
+        """Generates list of atom groups that satisfy select_list
+           strings
+
+        Parameters
+        ----------
+        frame : int
+            number of the current frame
+
+        Returns
+        -------
+        selections : list
+            list of atom groups
+        """
+        selections = []
+        for i in range(len(self.select_list)):
+            try:
+                selections.append(self.universe.select_atoms(self.select_list[i]))
+                assert selections[i].n_atoms > 0
+            except:
+                try:
+                    selections.pop(i)
+                except:
+                    pass
+                selections.append(None)
+        
+        return selections
+
+    def _init_pos_deque(self):
+        """Initializes lists necessary for MSD calculations at the
+           first restart.
+
+        Deques with maxlen=len_msd are used to conserve memory.
+
+        Returns:
+        --------
+        pos : list
+            contains position information, organized:
+            selection X frame X atom X (x, y, z)
+        set_list : list
+            contains set of atoms which satisfy each selection at each
+            frame, organized:
+            selection X frame X set of atom ids
+        dict_list : list
+            contains dictionaries which map atom ids to corresponding
+            pos index, organized:
+            selection X frame X dict
+        """
+
+        n_sel = len(self.select_list)
+        
+        # dictionaries with key: (id) and value: (array index)
+        # selection X frame X dictionary
+        dict_list = [[dict() for j in range(self.len_msd)]
+                     for i in range(n_sel)]
+        # atom ids which satisfy the selection at each frame
+        set_list = [[set() for j in range(self.len_msd)]
+                    for i in range(n_sel)]
+
+        # pre-allocate position array
+        pos = [[np.array([]) for j in range(self.len_msd)]
+               for i in range(n_sel)]
+
+        print("Populating deque...")
+        for ts in self.universe.trajectory[self.t0:self.t0
+                                           + self.len_msd]:
+            this_frame = ts.frame
+            selections = self._select(this_frame)
```

Comment:
coordinates of selections are automatically updated. You can just do this at the beginning of the loop. In fact you can to this at the beginning of the function and then set `n_sel = len(selections)`.